### ISSUE TITLE
Crash during the 'Checking helper files' window

### ISSUE THREAD

**Issue by pettrak** [aclist/dztui#112]
Running _dzgui.sh_ from the terminal I get this:

![image](https://github.com/aclist/dztui/assets/52935906/68713dec-516b-4365-aaca-4df1721605d0)

I don't receive the crash during the first start, but after I have updated and restarted DZGUI, it always crashes during the 'Checking helper files' window.

_Distro: OpenSUSE Tumbleweed
Desktop environment: KDE Plasma 6
Display server: Wayland_

**Comment by aclist** [comment]
Did this change occur after a system update?

This message means that when trying to initialize locales based on environment variables, it is finding a locale value that is unset on the system.

What is the output of `locale -a`?

Do you have the package `glibc-locale` installed?

If you export the LC_ALL environment variable to one of the ones printed in the list from `locale -a` above before running the application, does it work?

e.g. 

```
export LC_ALL=en_US.UTF-8
./dzgui.sh
```

**Comment by aclist** [comment]
See also https://superuser.com/questions/1838322/why-is-opensuse-tw-complaining-cant-set-the-locale-make-sure-lc-and-lang-a

Make sure your locales are reconfigured and up to date, and check if you haven't overriden locale preferences to a non-existent value somewhere in your shell settings.

**Comment by pettrak** [comment]
I fixed it by going to the graphical KDE System Settings page and changed everything to "sv_SE.UTF-8". I initially wanted to set the language to US English (en_US.UTF-8) and rest to Swedish since I much rather want to use the metric system, our date and time, currency, etc. It couldn't set a "LC_ALL=" since everything has to be of the of very same locale.

**Comment by aclist** [comment]
I think there are a couple of things going on here. This is going to be a bit long, but please read to the end.

For internationalization purposes, the default locale is strictly used to set numeric preferences like thousands separators and decimal notation. DZGUI currently has no alternate translations at the language level, so this is only for consistency in numeric values shown in tables like the server browser.

I am not sure if openSUSE is behaving in an unspecified way when it comes to setting the default locale. Even if using mixed settings, such as one "base" language for the language and then one "region" for numbers, this should still populate a default value that is readable through LC_ALL.

I found some hints in [this openSUSE forum thread](https://forums.opensuse.org/t/how-do-i-set-locale/127416/54), but nothing definitive. Warning: long and possibly out of date.

If I understood correctly, your initial setup consisted of en_US.UTF-8 as the default and sv_SE.UTF-8 as the region. But was this a recent change? Was this set through system files, environment variables, or a system settings dialog? Was there a point in time where you previously used a monolingual setup and were able to run DZGUI? I am just trying to figure out what caused LC_ALL to become unset or set to an "illegal" value, and whether or not mixed locales are actually the cause of the issue, or just something peripheral.

The reason I ask is because if you look at the terminal output above, when zenity was called, it actually complains about the default locale being unsupported: `Locale not supported by the C library`. zenity is not my program; it's the tool used to spawn the "Checking helper files" dialog. So even before DZGUI kicked in, zenity was trying to check the system's default locale (through LC_ALL, the standard convention) and came up empty or found something invalid there, so it defaulted to the C locale.

Can you reproduce this crash after explicitly setting the language preference to en_US.UTF-8 and the region to sv_SE.UTF-8? Try going back into the settings and seeing if explicitly appling this change will make the setting sticky now.

Some further explanation excerpted from [here](https://docs.python.org/3/library/locale.html):

> Initially, when a program is started, the locale is the C locale, no matter what the user’s preferred locale is. There is one exception: the [LC_CTYPE](https://docs.python.org/3/library/locale.html#locale.LC_CTYPE) category is changed at startup to set the current locale encoding to the user’s preferred locale encoding. The program must explicitly say that it wants the user’s preferred locale settings for other categories by calling setlocale(LC_ALL, '').
> 
> ```
> import locale
> locale.setlocale(locale.LC_ALL, '')
> ```
> 
> This sets the locale for all categories to the user’s default setting (typically specified in the LANG environment variable). If the locale is not changed thereafter, using multithreading should not cause problems.
> 
> According to POSIX, a program which has not called setlocale(LC_ALL, '') runs using the portable 'C' locale. Calling setlocale(LC_ALL, '') lets it use the default locale as defined by the LANG variable. Since we do not want to interfere with the current locale setting we thus emulate the behavior in the way described above.

As you can see, this is the POSIX-compliant method of inheriting the default system locale, rather than trying to access atomic locale values, which may be unreliable on some systems.

Even if you had used a mixed setup with en_US.UTF-8 and sv_SE.UTF-8, it should have caught at least one of the two and defaulted to that. This might be undesirable if you actually want metric notation but the base language is English, but it should not come up with an undefined locale as the default, which means the problem is somewhere else: somehow, the default locale was unset or malformed. Usually this means that the locale package for the chosen language is not installed on the system, so it's been set to something in name only. So if you actually want to use English as the locale default but don't install the necessary internationalization packages (should usually be installed to /usr/share/locale), various applications won't be able to take advantage of this setting.

So as a test, I would try different combinations of languages (EN base, SV region, or SV base, EN region) and seeing whether this causes a crash again or not. It shouldn't, provided you actually have those languages "installed" and viewable by the system. (This functionality is distribution specific)

Now, the next issue is that it seems you want SV numeric display regardless of the default language. Putting aside the issue of crashes at startup, even if you had successfully set EN as the default and SV as the region, since we are checking the default system language (for POSIX compatibility), as long as you set English, you are not going to get SV-style numeric notation.

For this case, what it seems you actually want is to set the locale at runtime for the context the application runs in:

`LC_ALL=sv_SE.UTF-8 ./dzgui.sh`

This will ensure numeric notation is in SV style solely for the DZGUI application. Provided SV language is installed, DZGUI will respect this value and inherit numeric preferences. You could wrap the above command in a shortcut or shell alias for convenience. This is not the root cause of the problem, but it provides a way of having custom numeric notation that differs from the system base language.

So, wrapping up:
1. If you have set mixed locales but the application crashes, it's probably an issue on the OS side due to a missing or malformed locale
2. If you have a regional numeric preference that's different from the base language, this is fine at the OS level, but since some distributions don't provide this functionality, for general compatibility, DZGUI will always look to the system's base language and set numerical preferences based on that. So if you want to override this behavior, pass a variable directly to it.

**Comment by aclist** [comment]
As a side note, when changing locales, you may have to start a new login session, as the window manager may still be using the old locale.

**Comment by pettrak** [comment]
> If I understood correctly, your initial setup consisted of en_US.UTF-8 as the default and sv_SE.UTF-8 as the region. But was this a recent change? 

This was initially set during the installation process of OpenSUSE while running on a boot media. Though by checking the KDE System Settings at Region & Language, I still saw some region settings using US standards, so I changed it. Though while searching for sv_SE.UTF-8 I saw other options from the same region (being "swedish (Sweden)") which didn't contain anything somehow. Must have accidentally chose a faulty one, causing the ``LC_ALL=`` to not be set. 

Going to YaST2's system settings I changed the system settings to Swedish and English as second option. Which probably installed the proper and new "sv_SE.UTF-8"-locales so it could properly work once I applied them on the Region & Language-page and I no longer saw the empty "swedish (Sweden)"-options. I also recently tried to change the language to "American English" (Strangely by typing ``locale`` in the terminal I get ``LANG=en_DK.UTF-8`` somehow) through KDE System Settings and kept the rest sv_SE.UTF-8 and DZGUI was still working.  

In conclusion. It must have been some faulty system setup by OpenSUSE. Since I didn't receive this issue while I have done similar configurations I mentioned earlier on Fedora KDE and didn't have nor ever had this issue. Maybe YaST and KDE doesn't communicate too well.

**Comment by pettrak** [comment]
> Are we good to close this ticket now?

We are good to close it now. Thanks for the help and contribution of the matter!
